### ISSUE TITLE
What to do with a big case statement?

### ISSUE THREAD

**Issue by lucasqueiroz** [rubocop-hq/ruby-style-guide#702]
Hello, I have the following problem:
I have a generator that calls a specific class for the param in the method, and a case/when with each code. But there is too many lines in the method (and rubocop is complaining). I'd like to know if there's a better way to do this...
My code so maybe it can clarify my explanation:
```ruby
def bank(bank_number)
      case bank_number
      when BRGenerator::BancoDoBrasil.CODE
        BRGenerator::BancoDoBrasil.bank
      when BRGenerator::Bradesco.CODE
        BRGenerator::Bradesco.bank
      when BRGenerator::Citibank.CODE
        BRGenerator::Citibank.bank
      when BRGenerator::HSBC.CODE
        BRGenerator::HSBC.bank
      when BRGenerator::Itau.CODE
        BRGenerator::Itau.bank
      when BRGenerator::Santander.CODE
        BRGenerator::Santander.bank
      else
        raise BRGenerator::InvalidBankError, "Invalid bank: #{bank_number}"
      end
    end
```
(Even fixing that with `when ... then ...`, it says the cyclomatic complexity is too high.)

**Comment by kolobock** [comment]
I would implement the sample code following above instead
```
BR_GENERATORS = %W(BancoDoBrasil Bradesco Citibank HSBC Itau Santander).freeze

def bank(bank_number)
  generator_name = BR_GENERATORS.detect {|brg| "BRGenerator::#{brg}".constantize.CODE == bank_number}

  raise BRGenerator::InvalidBankError, "Invalid bank: #{bank_number}" unless generator_name

  "BRGenerator::#{brg}".constantize.bank
end
```

**Comment by lucasqueiroz** [comment]
@kolobock thanks for the reply. That's a good approach to it. I also thought about creating a YML file, parsing it and saving to a hash inside the class (with the help of a loader class that I already use for other stuff), and then getting the class name from the hash. Now with your comment I'm not sure which one I'm going to use...
But thanks!

**Comment by mrbroll** [comment]
While @kolobock's sample will fix the method length issue, I don't think metaprogramming is the answer. You should never sacrifice the readability of your code in an effort to satisfy your linter. That defeats the purpose of using the linter in the first place.

@lucasqueiroz without more knowledge of `BRGenerator::<bank>` and the design decisions behind it, your case statement looks more like a key-value lookup than branching logic, so I would say you should do something like create a hash of `bank_number` to `bank`, and do a simple lookup in your method.

I have run into issues with the line length metric in methods that use case statements. Per the [documentation](https://rubocop.readthedocs.io/en/latest/cops_metrics/#metricsmethodlength) the default is 10, which effectively limits how many cases you can handle. I would say if you are using a large enough case statement, and it is idiomatic, then it's an exception to the rule and you should ignore the length metric for the method in question.
